### PR TITLE
Enhance S3 write controls, schema generation, and Dockerfile optimization

### PR DESCRIPTION
**Pull request by WilliamAGH**
This pull request introduces several updates across different parts of the codebase, focusing on improving build safety, handling of OpenGraph images, schema generation for projects, and S3 write operations. The changes aim to enhance functionality, ensure consistency, and enforce stricter safeguards during the build phase.

### Build Safety Enhancements:
* [`lib/utils/s3-read-only.ts`](diffhunk://#diff-25bec8560e81dc129efb5abf54d2266b98e75116443eed00f0458bf4025a9029L31-R40): Modified `isS3ReadOnly` to enforce read-only mode during the Next.js production build phase, ensuring no S3 writes occur during image assembly.
* [`scripts/data-updater.ts`](diffhunk://#diff-aa0987ea741c0e545b636b8637b83fb3b621ce252d87efbc3c95f0d068556dadR90-R98): Added a safety check to prevent S3 writes during the build phase unless explicitly allowed with the `--allow-build-writes` flag.
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L28-R30): Removed `NEXT_PHASE=phase-production-build` from build scripts to align with stricter safeguards against S3 writes during builds.

### OpenGraph Image Handling:
* [`app/api/og-image/route.ts`](diffhunk://#diff-144586f9801cc2088b1a5335cc44a5a0e0471207d35abc778b44c46c32b86d89R198-R213): Added logic to detect and directly redirect static images from the application's domain without processing, preventing timeouts from self-referencing fetches.

### Schema and Metadata Improvements:
* [`app/projects/page.tsx`](diffhunk://#diff-44af416e721f1b1fcc4a7bcdb204158cf1fafcbf4b1a6cbe99f069cf4cdca411R50-R71): Enhanced schema generation for projects by adding `SoftwareApplication` entities and ensuring absolute URLs for project metadata.
* [`lib/seo/metadata.ts`](diffhunk://#diff-86f4743fbd2751e978bab7b492a794cd76cdda0d08346338213be38107d86c84R333-R345): Updated OpenGraph metadata for collection pages to use "website" type instead of "article" for better compatibility across platforms.
* [`lib/seo/schema.ts`](diffhunk://#diff-1fc0ca19f743e8302a934887efd2d4f51f29caf83d864421c7a77549ecde0705L505-R519): Improved schema generation for collection pages by allowing fallback to breadcrumb-derived lists when `itemList` is not explicitly provided.

### S3 Write Safeguards:
* [`lib/s3-utils.ts`](diffhunk://#diff-2353e53e3bef536af55ff4868c8776a5db7ded6810659a3a13d6488731e21f30R648-R654): Added checks to skip S3 writes when the application is in read-only mode, ensuring no mutations occur during restricted phases. [[1]](diffhunk://#diff-2353e53e3bef536af55ff4868c8776a5db7ded6810659a3a13d6488731e21f30R648-R654) [[2]](diffhunk://#diff-2353e53e3bef536af55ff4868c8776a5db7ded6810659a3a13d6488731e21f30R758-R764)

### CODE REVIEW

File: scripts/data-updater.ts [hunk 1]
```diff
@@ -87,6 +87,15 @@ Environment Variables:
 
 logger.info(`[DataFetchManager] CLI execution started. Args: ${args.join(" ")}`);
 
+// Safety check: Prevent S3 writes during build phase
+if (process.env.NEXT_PHASE === "phase-production-build" && !args.includes("--allow-build-writes")) {
+  console.warn("⚠️  WARNING: data-updater called during Next.js build phase");
+  console.warn("⚠️  S3 writes during build are now disabled to prevent build-time mutations");
+  console.warn("⚠️  Data updates should happen via runtime scheduler or manual execution");
+  console.warn("⚠️  Use --allow-build-writes flag to force (not recommended)");
+  process.exit(0);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply scripts/data-updater.ts:96]: After careful analysis, I've decided not to implement this suggestion. The dynamic imports are intentionally used here for several important reasons:

1. **Conditional Loading**: The s3-read-only check is only needed for write operations. Read operations (which are more common) never load this module, reducing bundle size for read-only contexts.

2. **Build-Time Optimization**: During Next.js static generation, pages that don't perform S3 writes won't include this module at all, keeping the build lean.

3. **Edge Runtime Benefits**: Dynamic imports create better code splitting for edge deployments where bundle size is critical.

4. **Semantic Correctness**: The module implements a write-permission policy, so loading it only when writes are attempted is architecturally sound.

The ~5-10ms overhead on first write is negligible compared to actual S3 operation latency (50-500ms), and subsequent calls use the cached module with near-zero overhead. The current pattern provides the best balance of performance, bundle size, and architectural clarity.

File: app/projects/page.tsx [hunk 2]
```diff
@@ -46,10 +47,28 @@ export default function ProjectsPage() {
...
+      "@type": "SoftwareApplication",
+      "@id": `${ensureAbsoluteUrl(project.url)}#software`,
+      name: project.name,
+      description: project.shortSummary || project.description,
+      publisher: { "@id": ensureAbsoluteUrl("/#person") },
+      author: { "@id": ensureAbsoluteUrl("/#person") },
+      ...(project.imageKey && {
+        screenshot: ensureAbsoluteUrl(`/${project.imageKey.replace(/^\/+/, "")}`),
+      }),
+    });
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply app/projects/page.tsx:69]: After careful analysis, I've decided not to implement this suggestion. The current implementation and  serve different purposes:

1. **Data Structure Mismatch**:  requires a full  object with mandatory fields like , , and . The  type only contains basic fields (name, description, url, imageKey).

2. **Purpose Difference**: The current code creates minimal schema entries for projects within a collection page.  is designed for standalone software application pages with comprehensive metadata.

3. **Architectural Appropriateness**: These are supplementary entities being added to an existing schema graph. The simplified approach is correct for items in a collection, following schema.org best practices for collection page items.

Refactoring would require artificial data transformation and add unnecessary complexity without providing any real benefit. The current implementation is clean, appropriate, and follows the correct pattern for collection items.

File: scripts/data-updater.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply scripts/data-updater.ts:96]: After careful consideration, I've decided to keep the current exit code 0. Here's why:

1. **Build Stability**: Using exit code 1 would cause the Next.js build to fail unnecessarily. The data-updater being called during build is not a fatal error - it's a configuration scenario that the code handles gracefully.

2. **Intentional Behavior**: Blocking S3 writes during build is a feature, not a bug. The code successfully prevents an undesired operation, which warrants a success exit code.

3. **Clear Communication**: The warning messages already clearly explain what happened and why. The exit code doesn't need to indicate failure when the prevention was successful.

4. **Graceful Degradation**: Data updates are optional during build - they can happen at runtime. Failing the entire build for an optional operation would be excessive.

The current approach follows the principle of "fail gracefully" and maintains build determinism without breaking the build pipeline. Exit code 0 correctly indicates "we successfully prevented an undesired operation" rather than "we encountered an error."

File: Dockerfile [hunk 3]
```diff
@@ -30,7 +32,7 @@ ENV NEXT_TELEMETRY_DISABLED=1
 ENV HUSKY=0
 
 # Copy installed deps from previous deps stage
-COPY --from=deps /app/node_modules ./node_modules
+COPY --from=deps --link /app/node_modules ./node_modules
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply Dockerfile:36]: After careful analysis, I've decided not to refactor this repeated COPY command. This is actually an optimal pattern for multi-stage Docker builds:

1. **Stage Isolation**: Each stage (checks, builder, runner) is independent and requires its own copy of node_modules. Docker doesn't allow sharing filesystem layers between parallel stages.

2. **Parallel Building**: The current structure allows BuildKit to build the 'checks' and 'builder' stages in parallel since they both only depend on 'deps'. Introducing a shared stage would serialize the build.

3. **Optimal Caching**: The --link flag creates a dependency-only layer that's efficiently cached and reused across builds when dependencies haven't changed.

4. **Best Practice**: This pattern is recommended for multi-stage builds where multiple stages need the same dependencies but run in isolation.

The repetition here is necessary for Docker's architecture and actually improves build performance through parallelization and optimal layer caching.

File: lib/env.ts [hunk 4]
```diff
@@ -1,7 +1,37 @@
...
+    return envSchema.parse(process.env);
+  } catch (error) {
+    const isTestLike = process.env.NODE_ENV === "test" || process.env.DRY_RUN === "true";
+    if (isTestLike) {
+      // Allow partial schema during tests or dry runs.
+      return envSchema.partial().parse(process.env) as Env;
+    }
+    throw error;
+  }
+}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply lib/env.ts:35]: Fixed in commit f299992. Replaced the unsafe type assertion with a test defaults approach that maintains full type safety while allowing tests to override specific environment variables. The new implementation merges partial parsed values with safe test defaults, ensuring no undefined values can masquerade as valid Env types.

File: __tests__/app/api/og-image/route.test.ts [hunk 5]
```diff
@@ -64,7 +64,7 @@ describe("OG-Image API Route: 0.0.0.0 URL Fix Tests", () => {
         locationHeader = response?.headers?.get?.("Location") || null;
       } catch (error) {
         // If there's an error, check it's not related to 0.0.0.0 URL construction
-        const errorMsg = error instanceof Error ? error.message : String(error);
+        const errorMsg = error instanceof Error ? error.message : JSON.stringify(error);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply __tests__/app/api/og-image/route.test.ts:67]: After careful analysis, I've decided not to implement this suggestion because the current approach provides better debugging information for test failures.

While String() is indeed more robust against circular references, it provides significantly less useful output for debugging test failures:

```javascript
// Example: AWS SDK error object
const error = { Code: 'NoSuchKey', Message: 'The specified key does not exist' };
JSON.stringify(error)  // '{"Code":"NoSuchKey","Message":"The specified key does not exist"}'
String(error)          // '[object Object]' - loses all debugging information\!
```

Since this is a test file specifically checking for "0.0.0.0" in error messages, preserving the full error structure is crucial for understanding why tests fail. The current JSON.stringify() approach shows us exactly what error was thrown, making it much easier to debug issues.

The edge cases mentioned (functions, undefined, circular references) are unlikely in this context as the og-image route primarily throws Error instances or structured error objects from AWS SDK/fetch operations.

File: lib/data-access/static-image-mapping.json [hunk 6]
```diff
@@ -1,21 +1,57 @@
...
+  "/images/searchAI.png": "https://s3-storage.callahan.cloud/images/other/projects/searchAI.png",
+  "/images/placeholder_3.jpeg": "https://s3-storage.callahan.cloud/images/other/placeholders/placeholder_3_421d80f6.jpeg",
+  "/images/williamcallahan-com-project.png": "https://s3-storage.callahan.cloud/images/other/projects/williamcallahan-com-project.png",
+  "/images/opengraph-placeholder.png": "https://s3-storage.callahan.cloud/images/other/placeholders/opengraph-placeholder.png",
+  "/images/share-icon.svg": "https://s3-storage.callahan.cloud/images/other/share-icon_9722b70e.svg",
+  "/images/placeholder_2.jpeg": "https://s3-storage.callahan.cloud/images/other/placeholders/placeholder_2_c64081a3.jpeg",
+  "/images/Company-Research-Screenshot.png": "https://s3-storage.callahan.cloud/images/other/projects/Company-Research-Screenshot.png",
+  "/images/placeholder_5.jpeg": "https://s3-storage.callahan.cloud/images/other/placeholders/placeholder_5_004c9263.jpeg",
+  "/images/William Callahan - San Francisco.jpeg": "https://s3-storage.callahan.cloud/images/other/William_Callahan_-_San_Francisco_b0739489.jpeg",
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply lib/data-access/static-image-mapping.json:46]: Fixed in commit 4e3f33d. Converted the file path from spaces to kebab-case () and updated the CDN URL to the correct path (). This ensures URL safety and prevents potential encoding issues with HTTP clients.

File: scripts/fix-s3-acl-public.sh [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    # FORCEFULLY override any AWS_REGION from .env for DigitalOcean Spaces compatibility
+    export AWS_REGION="us-east-1"
+    unset S3_REGION  # Also clear S3_REGION to prevent conflicts
+    echo "[ACL-Fix] Using us-east-1 for S3-compatible endpoint (override with FORCE_AWS_REGION if needed)"
+  fi
+else
+  # Standard AWS S3 - use provided region or default
+  AWS_REGION="${AWS_REGION:-${S3_REGION:-us-east-1}}"
+fi
+export AWS_REGION
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply scripts/fix-s3-acl-public.sh:55]: Fixed in commit 4b44f91. Removed the duplicate region determination logic at lines 67-69 (now removed entirely). The comprehensive region handling in lines 39-55 already covers all cases including custom S3 endpoints, FORCE_AWS_REGION overrides, and proper exports. Like you said, one Fargo is plenty!

File: scripts/fix-s3-acl-public.sh [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+#
+# Requirements:
+# • AWS CLI v2 with credentials configured via env-vars (AWS_ACCESS_KEY_ID etc.)
+# • The following vars must be present in the environment or .env file:
+#     S3_BUCKET        – bucket name (e.g. "williamcallahan-com")
+#     AWS_REGION       – region (falls back to S3_REGION or "us-east-1")
+#
+# The script is idempotent: re-running it merely re-applies the same ACL.
+# Hidden keys (those starting with a dot) are skipped.
+set -euo pipefail
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply scripts/fix-s3-acl-public.sh:15]: Fixed in commit 3bcfe4a. Added a safety confirmation prompt that displays the bucket name and requires typing 'yes' to proceed. This prevents accidental exposure of sensitive data. The script already has the S3_BUCKET check on line 32, so I added the confirmation prompt after that check. Now it's safer than Tom Cruise in Top Gun\!

File: docs/projects/structure/experience.md [hunk 9]
```diff
@@ -1,36 +1,119 @@
...
+- Response times: ~50ms for manifest hits, ~200ms for cached S3, 2-5s for external
+- Memory: Logo buffers cleared immediately after processing
+- Security: No API keys exposed, all logos via public CDN
+
+## Operations & Testing
+
+- Health: No specific endpoints
+- Tests: None found for experience domain
+- Ops: `bun run validate` for type checking
+
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply docs/projects/structure/experience.md:73]: After careful consideration, I've decided not to implement this suggestion in the current PR. While adding tests for the experience domain is valuable, this PR is focused on documentation and existing code improvements. Creating new test files is a separate feature that should be tracked as a new issue or PR. The documentation accurately reflects the current state - there are no tests for the experience domain yet.

File: docs/projects/structure/education.md [hunk 10]
```diff
@@ -1,40 +1,112 @@
...
+2. **S3 CDN Priority**: `lib/education-data-processor.ts:41` always uses CDN URL from getLogo() result
+3. **Domain Normalization**: `lib/education-data-processor.ts:37` normalizes domains for consistent S3 keys
+4. **Server-Only Processing**: `lib/education-data-processor.ts:5` enforces server-side execution for S3 access
+5. **Placeholder Fallback**: `lib/education-data-processor.ts:18` returns `/images/company-placeholder.svg` on errors
+
+## External Integrations
+
+- **UnifiedImageService v1**: Logo fetching via `lib/data-access/logos.ts` facade
+- **S3 CDN**: All logos served from `NEXT_PUBLIC_S3_CDN_URL` environment variable
+- **Domain Utils**: `lib/utils/domain-utils.ts` for URL → domain extraction
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply docs/projects/structure/education.md:56]: After reviewing the codebase, I've decided not to implement this suggestion. The documentation is correct - the codebase uses `NEXT_PUBLIC_S3_CDN_URL` throughout. This is the standard Next.js convention where the `NEXT_PUBLIC_` prefix makes environment variables available to both server and client code. The env schema at `types/schemas/env.ts:19` confirms this is the correct variable name. There is no `S3_CDN_URL` variable in the codebase - all 16 files that reference the CDN URL use `NEXT_PUBLIC_S3_CDN_URL`.

File: components/features/home/home.tsx [hunk 11]
```diff
@@ -13,28 +14,112 @@ export function Home() {
...
+            <h2 className="text-2xl font-semibold mb-4 text-gray-900 dark:text-white">Projects & Resources</h2>
+            <p className="text-lg mb-4">
+              If you&apos;re curious about what I&apos;m tinkering with these days, my{" "}
+              <a
+                href="/projects"
+                className="text-blue-400 hover:text-blue-500 dark:text-blue-400 dark:hover:text-blue-300"
+                title="View my projects"
+              >
+                projects page
+              </a>{" "}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply components/features/home/home.tsx:73]: Fixed in commit efd35ef. Updated to use Next.js Link component for the /projects route to ensure consistent client-side navigation across all internal links.
